src/v/cluster/utils/notifications.h

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+#include "cluster/notification.h"
+#include "cluster/topic_configuration.h"
+#include "model/fundamental.h"
+
+#include <seastar/util/noncopyable_function.hh>
+
+namespace cluster {
+
+// Unified interface for partition level notifications.
+// Supports notifications for partition leadership changes,
+// partition management (creation, deletion), and partition properties changes.
+class notifications {
+public:
+    // bool is true if the ntp local to this shard is the leader
+    struct partition_state {
+        partition_state(
+          model::ntp ntp,
+          model::term_id term,
+          bool is_leader,
+          std::optional<cluster::topic_configuration> topic_cfg)
+          : ntp(std::move(ntp))
+          , term(term)
+          , is_leader(is_leader)
+          , topic_cfg(std::move(topic_cfg)) {}
+        model::ntp ntp;
+        model::term_id term;
+        bool is_leader;
+        std::optional<cluster::topic_configuration> topic_cfg;
+    };
+    using notification_cb_t = ss::noncopyable_function<void(partition_state)>;
+
+    notifications() = default;
+    notifications(const notifications&) = delete;
+    notifications(notifications&&) = delete;
+    notifications& operator=(const notifications&) = delete;
+    notifications& operator=(notifications&&) = delete;
+
+    virtual ~notifications() = default;
+
+    virtual notification_id_type
+      register_leadership_notification(notification_cb_t)
+      = 0;
+    virtual void unregister_leadership_notification(notification_id_type) = 0;
+
+    virtual notification_id_type
+    register_partition_manage_notification(const model::ns&, notification_cb_t)
+      = 0;
+    virtual void unregister_partition_manage_notification(notification_id_type)
+      = 0;
+
+    virtual notification_id_type register_partition_unmanage_notification(
+      const model::ns&, notification_cb_t)
+      = 0;
+    virtual void
+      unregister_partition_unmanage_notification(notification_id_type)
+      = 0;
```

Comment:
Do we ever care about the distinction between these 3 cases anywhere for this "follow leadership" use case? Can we just merge them?

Reply:
Thats a good point .. we could probably merge all of them into a single call.